### PR TITLE
deadlock: only observe valid region

### PR DESCRIPTION
**Pull request by youjiali1995**
Signed-off-by: youjiali1995 <zlwgx1023@gmail.com>

###  What have you changed?
A new peer's region info is not complete until applying snapshot which causes the leader of the deadlock detector stepping down. Only observing valid region can fix it.

###  What is the type of the changes?
- Bugfix (a change which fixes an issue)

###  How is the PR tested?
- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)

###  Does this PR affect documentation (docs) or should it be mentioned in the release notes?
Yes

###  Does this PR affect `tidb-ansible`?
No

###  Any examples? (optional)
Without the fix, the new added test will panic when a new peer is added to the store with the leader of deadlock detector.
```
---- server::lock_manager::test_detect_deadlock_when_shuffle_region stdout ----
thread 'server::lock_manager::test_detect_deadlock_when_shuffle_region' panicked at 'assertion failed: resp.errors[0].has_deadlock()', tests/integrations/server/lock_manager.rs:61:5
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.29/src/backtrace/libunwind.rs:88
   1: backtrace::backtrace::trace_unsynchronized
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.29/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:47
   3: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:36
   4: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:198
   5: std::panicking::default_hook
             at src/libstd/panicking.rs:209
   6: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:475
   7: std::panicking::begin_panic
   8: integrations::server::lock_manager::must_deadlock
             at tests/integrations/server/lock_manager.rs:61
   9: integrations::server::lock_manager::test_detect_deadlock_when_shuffle_region
             at tests/integrations/server/lock_manager.rs:169
  10: integrations::server::lock_manager::test_detect_deadlock_when_shuffle_region::{{closure}}
             at tests/integrations/server/lock_manager.rs:117
  11: core::ops::function::FnOnce::call_once
             at /rustc/0e4a56b4b04ea98bb16caada30cb2418dd06e250/src/libcore/ops/function.rs:231
  12: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
             at /rustc/0e4a56b4b04ea98bb16caada30cb2418dd06e250/src/liballoc/boxed.rs:746
  13: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:82
  14: std::panicking::try
             at /rustc/0e4a56b4b04ea98bb16caada30cb2418dd06e250/src/libstd/panicking.rs:273
  15: std::panic::catch_unwind
             at /rustc/0e4a56b4b04ea98bb16caada30cb2418dd06e250/src/libstd/panic.rs:388
  16: test::run_test::run_test_inner::{{closure}}
             at src/libtest/lib.rs:1466
```

### CODE REVIEW

**Comment by youjiali1995** [comment]
@zhangjinpeng1987 @BusyJay @AndreMouche @MyonKeminta @sticnarf PTAL

File: src/server/lock_manager/deadlock.rs [hunk 1]
```diff
@@ -224,7 +224,8 @@ const LEADER_KEY: &[u8] = b"";
 
 /// Returns true if the region containing the LEADER_KEY.
 fn is_leader_region(region: &'_ Region) -> bool {
-    region.get_start_key() <= LEADER_KEY
+    !region.get_peers().is_empty()
+        && region.get_start_key() <= LEADER_KEY
         && (region.get_end_key().is_empty() || LEADER_KEY < region.get_end_key())
```
Comment (AndreMouche) [inline src/server/lock_manager/deadlock.rs]: Could we remove L228?

Comment (5kbpers) [inline src/server/lock_manager/deadlock.rs]: Is this condition necessary? I guess it will be always 'true'.

Comment (youjiali1995) [reply src/server/lock_manager/deadlock.rs]: It's not necessary, but it's an idiom.

Comment (youjiali1995) [reply src/server/lock_manager/deadlock.rs]: It's not necessary, but it's an idiom to check key in range.

File: src/server/lock_manager/deadlock.rs [hunk 2]
```diff
@@ -224,7 +224,8 @@ const LEADER_KEY: &[u8] = b"";
 
 /// Returns true if the region containing the LEADER_KEY.
 fn is_leader_region(region: &'_ Region) -> bool {
-    region.get_start_key() <= LEADER_KEY
+    !region.get_peers().is_empty()
```
Comment (Connor1996) [inline src/server/lock_manager/deadlock.rs]: you should add a comment here

File: components/pd_client/src/lib.rs [hunk 3]
```diff
@@ -149,9 +149,8 @@ pub trait PdClient: Send + Sync {
     }
 
     /// Gets Region info which the key belongs to.
-    fn get_region_info(&self, key: &[u8]) -> Result<RegionInfo> {
-        self.get_region(key)
-            .map(|region| RegionInfo::new(region, None))
+    fn get_region_info(&self, _key: &[u8]) -> Result<RegionInfo> {
```
Comment (BusyJay) [inline components/pd_client/src/lib.rs]: Will the implement of pd_client be affected by the change?

Comment (youjiali1995) [reply components/pd_client/src/lib.rs]: No. It's implemented by all pd_clients except the dummy pd_client.

File: src/server/lock_manager/deadlock.rs (hunk 2 above)
Comment (youjiali1995) [reply src/server/lock_manager/deadlock.rs]: Done

**Comment by youjiali1995** [comment]
'raftstore::test_split_region::test_incompatible_server_auto_split_region' failed.
/test

File: src/server/lock_manager/deadlock.rs [hunk 4]
```diff
@@ -224,7 +224,12 @@ const LEADER_KEY: &[u8] = b"";
 
 /// Returns true if the region containing the LEADER_KEY.
 fn is_leader_region(region: &'_ Region) -> bool {
-    region.get_start_key() <= LEADER_KEY
+    // The key range of a new created region is empty which misleads the leader
```
Comment (Connor1996) [inline src/server/lock_manager/deadlock.rs]:
```suggestion
    // The key range of an uninitialized peer of one region is empty which misleads the leader
```

Comment (youjiali1995) [reply src/server/lock_manager/deadlock.rs]: I will file a new PR to change it.

**Pull request merged**